rfcs/0004-leader-rotation.md

```diff
@@ -1,29 +1,23 @@
 # Leader Rotation
 
-The goal of this RFC is to define how leaders rotate in taking their role.
+The goal of this RFC is to define how fullnodes rotate in taking the leader role.
 
 ## Leader Schedule Generation
 
 Leader schedule is decided via a predefined seed.  The process is as follows:
 
-1. Periodically at a specific `PoH tick count` use the tick count (simple monotonically increasing counter) as a seed to a stable pseudo-random algorithm.
-2. At that height, example the bank for all the currently staked accounts and their assigned leader identities and weights.
+1. Periodically use the PoH tick height (a monotonically increasing counter) to seed a stable pseudo-random algorithm.
+2. At that height, sample the bank for all the currently staked accounts, their assigned leader identities and weights.
 3. Sort them by stake weight.
-4. Using the random seed select nodes weighted by stake to create a stake weighted ordering.
-5. This ordering becomes valid in `N` number of slots.
+4. Use the random seed select nodes weighted by stake to create a stake weighted ordering.
+5. This ordering becomes valid after a cluster-configured number of ticks.
 
 The seed that is selected is predictable but unbiasable.  There is no grinding attack to influence its outcome.  To reduce the likelihood of censorship of leaders in the network the set of **staked accounts** is computed over a large period.
 
-## Leader Rotation
+## Appending Entries
 
-* The leader is rotated every `T` PoH ticks (leader period), according to the leader schedule.  This amount of time as represented by the PoH ticks is called a slot.
+A leader schedule is split into *slots*, where each slot has a duration of `T` PoH ticks.
 
-Leader's transmit for a count of `T` PoH ticks.  When `T` is reached all the validators should switch to the next scheduled leader.  Leaders that transmit out of order can be ignored.
+A leader transmits entries during its slot.  After `T` ticks, all the validators switch to the next scheduled leader. Validators must ignore entries sent outside a leader's assigned slot.
 
-All `T` ticks must be observed from the current leader for that part of PoH to be accepted by the network.  If `T` ticks (and any intervening transactions) are not observed, the network optimistically fills in the `T` ticks, and continues with PoH from the next leader.  See [fork generation](0002-fork-generation.md).
-
-## Network Variables
-
-`N` - Number of slots before a new leader schedule becomes active after it is computed.
-
-`T` - Number of PoH ticks per leader slot.
+All `T` ticks must be observed from the current leader for that part of PoH to be accepted by the cluster.  If entries are not observed (leader is down) or entries are invalid (leader is buggy or malicious), a validator should fill the gap with empty entries and continue with PoH from the next leader.
```

Comment:
The new leader must transmit any ticks that would complete the PoH chain, i.e. any ticks that come after the last slot observed by the node and the node's slot.

Reply:
On second thought, the next leader doesn't need to produce those ticks. It *could*, but it'd consume network bandwidth and its blob would have a mismatch between PoH hash and signature. The validator would need to be smart enough to see that as a valid exception, and if it were that smart, why not make it smart in a different way - fill in the blank locally and before the next leader's first entry arrives? No delays and no bandwidth required.

Reply:
@garious which branch is the leader connected to?

Reply:
The next leader would point to the last empty entry. In every slot, the validator would produce an EntryTree with both heads, one from the expected leader and one with empty entries. After the first entry is received from the next leader, the validator would prune one fork and repeat the process.

Reply:
I worry you've already had this discussion and it never made it into an RFC?

Reply:
@garious leaders and validators don't have agreement on what is the `last empt slot`.

Reply:
What do you mean? The slot duration is a fixed number of PoH hashes. With no transactions to bias it, all nodes will conclude the same hash, which the next leader would build on.

Reply:
They don’t have agreement on the value.  Some have received valid data, some haven’t. Validators start a single PoH core from their last vote, until it is their slot to transmit.

Reply:
So let's say a leader can produce 2 entries in its slot. If the next leader only observed the first entry, I think there are 2 design options:

1. It could build an empty entry on top of the one observed entry, send that out, and have validators accept it even though the empty entry is signed by the wrong leader for that slot. I think this is what you and @rob-solana are proposing.
2. It can choose to build its own entries on a slot of *all* empty entries, effectively forcing validators that observed the valid entry to roll back. This was my understanding of how the system works with "virtual ticks". Is it possible the concept of "virtual ticks" was abandoned and I missed the memo?

Reply:
Please, the question about "virtual ticks", was that approach abandoned?

Reply:
@garious they are still virtual ticks since the ticks are the same if they are derived from the same slot regardless of which leader proposes them.